src/main/scala/pirate.internal/ParseTraversal.scala

```diff
@@ -26,11 +26,30 @@ object ParseTraversal {
   def errorMessageP[A](e: String): P[A] =
     errorP(ParseErrorMessage(e))
 
-  def exitP[A, B](p: Parse[B], o: Option[A]): P[A] =
-    hoistMaybe(o)
-
-  def hoistMaybe[A](o: Option[A]): P[A] =
-    o.map(a => P[A](_ => a.right)).getOrElse(errorP(ParseErrorNoMessage))
+  def exitP[A, B](p: Parse[B], o: ValidationNel[MissingParser, A]): P[A] =
+    hoistValidation(o)
+
+  def hoistValidation[A](o: ValidationNel[MissingParser, A]): P[A] =
+    o.fold(es =>
+      // If there's a missing branch, just show the help text.
+      if (es.toList.collect { case MissingOptions(m) => m }.length > 0)
```

Comment:
FWIW I try to avoid `collect`, as convenient as it is. I might be tempted to add a `fold` on `MissingParser`. /suggestion